content/articles/mock-a-datasource-for-react-using-json-server/index.md

```diff
@@ -0,0 +1,736 @@
+#### Full-text search
+e.g. `GET/posts?q=tomcat`. This request searches for the value '**tomcat**' in all the records stored. An example is a GET request on `http://www.albana.com/posts?q=tomcat`.
+#### Relationships
+e.g., the inclusion of child resources(`_embed`), the inclusion of parent resources(`_expand`), to get or create nested resources. An example is `http://www.albana.com/comments/1?_expand=post` that includes parent resource by the name 'post'.
+#### Database
+e.g. `GET\db`. This GET request brings up all database items of the specified database. An example is `http://www.albana.com/customers_db`.
+#### Homepage
+e.g. `GET/`. Using the `./public` path, one can serve the application files in the public folder. An example is `http://www.albana.com/`.
+
+More on JSON Server library will also be mentioned in the article.
+
+### Create a React app
+
+In this section, a React app will be created. The app will be a budget planning application to add items and remove them from a list.
+The app saves and retrieves data from a data source in a JSON object.
+
+#### New React app
+
+- Open the folder in which the application should be created and run this command:
+```bash
+npx create-react-app react-budget-tracker
+```
+
+#### New project structure
+- Create the following application folder structure:
+```bash
+.
+├── node_modules
+├── public
+├── src
+│   └── components
+│       ├── About.js
+│       ├── AddBudget.js
+│       ├── Budget.js
+│       ├── Budgets.js
+│       ├── Button.js
+│       ├── Footer.js
+│       └── Header.js
+│   ├── App.js
+│   ├── index.css
+│   ├── index.js
+│   └── reportWebVitals.js
+├── .eslintcache
+├── .gitignore
+├── db.json
+├── package.json
+├── package-lock.json
+├── README.md
+└── yarn.lock
+```
+
+#### Header component
+As the name suggests, the `Header.js` file contains the header parts.
+- In the `Header.js`, its code is as shown below:
+```js
+import PropTypes from 'prop-types'
+import { useLocation } from 'react-router-dom'
+import Button from './Button'
+
+const Header = ({ title, onAdd, showAdd }) => {
+  const location = useLocation()
+
+  /*The header has a title and a button that opens up a window to add the budget or close the window*/
+  return (
+          <header className='header'>
+
+            <!-- The header has a title -->
+            <h1>{title}</h1>
+            {location.pathname === '/' && (
+                    <Button
+                            color={showAdd ? 'red' : 'green'}
+                            text={showAdd ? 'Close' : 'Add'}
+                            onClick={onAdd}
+                    />
+            )}
+          </header>
+  )
```

Comment:
Are these formatted with 2 or 4 tab spaces each?

Possibly, you can choose 2

Reply:
ok